### PR TITLE
Visacheckout

### PR DESCRIPTION
**Pull request by aengusbates**
Notes:
- Refactoring UI to accommodate 7 payment types.
  - (UI design based on specs for EU Checkout product.)
- Checkout demo is now dissociated from the rest of the payment flows to mirror real world use cases.
- Adding Funds Transfer payment type.
- Added Visa Checkout. 
- Added Masterpass.

Local links:
  - https://0.0.0.0:5000/
  - https://0.0.0.0:5000/checkout

Deployed to new staging env for internal testing:
- Main: https://na-merchant-api-demo-staging.us-west-2.elasticbeanstalk.com/
- Visa Checkout: https://na-merchant-api-demo-staging.us-west-2.elasticbeanstalk.com/visa-checkout
- Checkout: https://na-merchant-api-demo-staging.us-west-2.elasticbeanstalk.com/checkout

### CODE REVIEW

**Comment by aengusbates** [comment]
Known bug. Redirect on clicking cancel on Interac isn't handled properly. This is a legacy bug (needs to be fixed, but not in this pull).

File: server/app/blueprints/funds_transfer.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    # first, submit batch file containing transaction
+    # second, get report to varify status of transaction
+
+    success = 'false'
+    batch_id = ''
+    trans_id = ''
+
+    amount = int(float(request.json.get('amount')) * 100) # convert dollars to cents
+    name = request.json.get('name')
+    accountNumber = request.json.get('accountNumber')
```
Comment (svenresch) [inline server/app/blueprints/funds_transfer.py]: 1.) Is transfer of this info, from a customers Web browser to a Merchants Server to the Bambora payments endpoint, in a non-tokenized manner really a very good idea?

2.) Is using the batch upload/processing API for this kind of singular on-demand request really a good idea? Me thinks not as it was setup for batch processing and not really for an OLTP type of system.

Comment (aengusbates) [reply server/app/blueprints/funds_transfer.py]: I'll remove the payment type for the moment.

1. This is how merchants currently collect banks data. Ideally, we would add it to Custom Checkout.

2. We're omitting the batching stage for simplicity - but we're also omitting the consent stage which may give the wrong idea about use case. The main thing was to demonstrate how to use the Batch Payment API and Batch Payment Report API - both are complex APIs.

**Comment by aengusbates** [comment]
Code related to Funds Transfer demo removed.

File: server/Makefile [hunk 2]
```diff
@@ -1,9 +1,9 @@
 REPO_BASE=056252067802.dkr.ecr.us-west-2.amazonaws.com
 APP_NAME=na-merchant-api-demo
 IMAGE_NAME=bambora/$(APP_NAME)
-VERSION=1.0.0
+VERSION=9.0.0
 EB_BUCKET=eb-docker-deploy
-EB_APP_ENV_NAME=$(APP_NAME)
+EB_APP_ENV_NAME=$(APP_NAME)-staging
```
Comment (svenresch) [inline server/Makefile]: Should this be left as staging?

File: server/app/blueprints/checkout.py [hunk 3]
```diff
@@ -5,30 +5,40 @@
...
+                '&ordName=' + name + '&ordPostalCode=' + postal +
+                '&ordProvince=BC&ordCountry=CA' +
+                '&approvedPage=' + term_url + '&declinedPage=' + term_url)
 
-    hash_data = 'merchant_id=' + merchant_id + '&trnAmount=' + amount + '&ordName=' + name + '&ordPostalCode=' + postal
-    hash_key = '204AA6A5-1BCD-489E-9FEA-C904EBFA'
     hash = hashlib.sha1(str(hash_data + hash_key).encode('utf-8')).hexdigest()
 
-    checkout_url = 'https://web.na.bambora.com/scripts/payment/payment.asp?'+ hash_data + "&hashValue=" + hash
+    checkout_url = '{}/scripts/payment/payment.asp?{}&hashValue={}'.format(settings.base_querystring_sandbox_url, hash_data, hash)
```
Comment (svenresch) [inline server/app/blueprints/checkout.py]: Do we not have a REST endpoint for this?

File: server/app/blueprints/masterpass.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    else:
+        feedback = {'success': False,
+                    'message': (params['messageId'][0] + ': ' + params['messageText'][0])}
+
+    return json.dumps(feedback), status
+
+
+@payments.route('/callback', methods=['GET'])
+def callback():
+    url = '{}/scripts/process_transaction.asp'.format(settings.base_querystring_sandbox_url)
```
Comment (svenresch) [inline server/app/blueprints/masterpass.py]: Do we not have a REST endpoint for this?

File: server/app/blueprints/visa_checkout.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import settings
+
+
+payments = Blueprint('visa-checkout', __name__,)
+
+
+@payments.route('/capture', methods=['POST'])
+def make_payment():
+
+    url = '{}/scripts/process_transaction.asp'.format(settings.base_querystring_sandbox_url)
```
Comment (svenresch) [inline server/app/blueprints/visa_checkout.py]: Do we not have a REST endpoint for this?

File: server/app/settings.py [hunk 6]
```diff
@@ -14,14 +14,29 @@
...
+base_querystring_url = 'https://web.na.bambora.com'
+base_querystring_sandbox_url = 'https://sandbox-web.na.bambora.com'
+
 # Bambora params needed for authentication include Merchant ID & API Passcode.
 # --> More info here: https://dev.na.bambora.com/docs/guides/merchant_quickstart/
 merchant_id = os.environ.get('MERCHANT_ID')
 api_passcode = os.environ.get('API_PASSCODE')
+report_api_passcode = os.environ.get('REPORT_API_PASSCODE')
+
+sandbox_merchant_id = os.environ.get('SANDBOX_MERCHANT_ID')
```
Comment (svenresch) [inline server/app/settings.py]: Maybe an explanation for what the regular/prod vars are versus the sandbox ones would be helpful. What is the sandbox?

File: server/app/templates/snippets/panel-visa-checkout.html [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    <img alt="Visa Checkout" class="v-button" role="button" data-locale="en_CA"
+    src="https://sandbox.secure.checkout.visa.com/wallet-services-web/xo/button.png"/>
+
+    <script type="text/javascript"
+      src="https://sandbox-assets.secure.checkout.visa.com/checkout-widget/resources/js/integration/v1/sdk.js">
+    </script>
+
+    <div class="test-cards">
+      <p>Visa Cheeckout is a digital wallet that allows consumers to pay with any payment card, on a website or app, with any device.</p>
```
Comment (svenresch) [inline server/app/templates/snippets/panel-visa-checkout.html]: Checkout is mis-spelled here

File: server/app/blueprints/checkout.py (hunk 3 above)
Comment (aengusbates) [reply server/app/blueprints/checkout.py]: Nope.

File: server/Makefile (hunk 2 above)
Comment (aengusbates) [reply server/Makefile]: Good catch. Local change. Will revert.

File: server/app/blueprints/masterpass.py (hunk 4 above)
Comment (aengusbates) [reply server/app/blueprints/masterpass.py]: Nope.

File: server/app/blueprints/visa_checkout.py (hunk 5 above)
Comment (aengusbates) [reply server/app/blueprints/visa_checkout.py]: Nope.

File: server/app/settings.py (hunk 6 above)
Comment (aengusbates) [reply server/app/settings.py]: That would be helpful. I'll add a comment.

File: server/app/templates/snippets/panel-visa-checkout.html (hunk 7 above)
Comment (aengusbates) [reply server/app/templates/snippets/panel-visa-checkout.html]: Cheeky Checkout ;)

**Pull request merged**